### PR TITLE
Refactor/scoped ddp training

### LINKED ISSUE #111: --start-at-iteration N prevents updating alignments after each iteration
When using the `--start-at-iteration` flag to continue from an interrupted iteration, using any value other than 0 appears to prevent alignments from being updated. The tilt-series alignment step takes significantly shorter time (on the order of minutes rather than hours) and the resultant xml files are identical to the previous iteration's. 

Command: `CUDA_VISIBLE_DEVICES=0,1,2,3 OMP_NUM_THREADS=1 MKL_NUM_THREADS=1 miss-alignment --config-file config.yaml --training-devices 0,1,2,3 --reconstruction-devices 0,0,1,1,2,2,3,3 --start-at-iteration 2 --ddp-timeout-hours 12`
<details><summary>config.yaml</summary>

```yaml
general:
  # MissAlignment iteratively trains models and realigns the tilt-series
  training_directory: missalignment  # path to directory for writing iteration output
  apply_ctf: True                # set to True if CTF estimates are available in .xml
  iteration_settings:             # defines all iterations to run (length determines number of iterations)
    # alignment modes: "global" (single pass), "anchoring" (iterative), "spline" (coarse-to-fine)
    - { downsample: 3, alignment: anchoring }
    - { downsample: 2, alignment: anchoring }
    - { downsample: 1, alignment: global }
    - { downsample: 1, alignment: global }
    - { downsample: 1, alignment: [3, 3] }  # local refinement with a 3x3 image warping grid
    - { downsample: 1, alignment: [3, 3] }
    - { downsample: 1, alignment: [3, 3] }
    - { downsample: 1, alignment: [3, 3] }
  seed: 45132

model_training:
  # used to initialize model weights
  model_architecture: 'default'
  model_checkpoint: null
  loss_margin: 0.5
  learning_rate: 1.0e-3
  # Set to zero to disable weight decay (i.e. the AdamW optimizer)
  weight_decay: 1.0e-4
  max_epochs_per_iteration: 30  # absolute maximum of stopping
  warmup_steps: 500
  multistep_lr_scheduler:
    milestones: [5, 15]
    gamma: 0.5                  # multiply learning rate by this value at the milestone epochs

data_loading:
  batch_size: 48                  # these value have been used as defaults:
  patch_size: 96                  # batch size 32 | reconstruction size 128 ^ 3
  steps_per_epoch: 1000           # this defines the size of an epoch

# this configures the shift generation for the contrastive training of MissAlignment
# all parameters are in units of pixels, thus shifts become larger when downsampling
shift_generation:
  trajectory_probability: .5      # these make bananas and birds
  trajectory_max_shift: 10.0      # trajectories fall somewhere between +/- f
  jitter_probability: .5
  jitter_max_std: 2.0             # maximum standard deviation of a normal distribution
  outlier_probability: .5
  outlier_max_shift: 20.0         # outlier can maximally go to this value
  fracture_probability: .5
  fracture_max_shift: 20.0       # specific strong shifts for high tilt images up to this value

tilt_series_alignment:
  patch_size: 96      # same as training patch size
  patch_overlap: 0.1  # tolerated overlap between patches used for optimizing the alignment
  batch_size: 48      # amount of patches simultaneously reconstructed in memory -> the more the merrier
```

</details>

### LINKED ISSUE #115: handle n_procs for preprocessing similar to alignment

### PR DESCRIPTION
**Pull request by McHaillet**
Core refactor — scoped DDP
  - Training phase now runs one mp.spawn worker per GPU, joined per macro-iteration; main process is single-process for alignment (no idle ranks/barriers); removed
  distributed_barrier/rank_zero_print/ddp_timeout_hours. 
  - Pinned LightningEnvironment so SLURM can't hijack the spawned ranks; extracted _set_ddp_env. Tests cover the launcher + SLURM-override.

  Parallelism consolidation
  - Added shared _parallel.run_device_pool (one process per GPU + work queue); migrated prepare-stacks, preprocessing, and alignment onto it; dropped the fixed n_processes knob.
  - Hardened spawned-process cleanup: reconstruction workers daemon=True, bounded/escalating teardown.
  - Capped TORCHINDUCTOR_COMPILE_THREADS to available_cpus // n_training_devices (cpuset/SLURM-aware).

  Resume (#111) fixes
  - Load prior iteration's checkpoint when resuming; training_model_path as single source of truth.
  - Removed pre-loop model.ckpt copy that crashed resume (SameFileError).
  - On resume, restore working XMLs from iter{N}/ (back up only at iter 0). Helpers _resolve_start_checkpoint/_sync_start_iteration_xmls unit-tested.

  Logging / output cleanup
  - MISS_ALIGNMENT_LOG_LEVEL env var (worker-aware); muted Lightning INFO banners (both lightning.pytorch + lightning.fabric); explicit TensorBoard logger; recon lifecycle → debug;
  silenced repeated seed/model-summary prints.
  - Progress bars always shown; device output grouped by macro-iteration phase.

  Docs
  - CLAUDE.md: macro-iteration/scoped-DDP rationale, stable multi-lab framing, trimmed global-overlap.
  - docs/usage.md: single-task SLURM, log level, compile-threads notes.

This makes PR #113 obsolete
closes #115 
closes #111

### CODE REVIEW

**Comment by McHaillet** [comment]
Tested on a multi-gpu system on the SHREC dataset. Everything is consistent, so merging

**Comment by CACNYC** [comment]
@McHaillet Great to see a new release! Compared to 0.1.6,  is release 0.1.7 a critical performance quality update or mainly performance speed update?

**Comment by McHaillet** [comment]
Hey! There is a training speed improvement from the suggestion from amineuron, mainly some knobs from pytorch lightning that give a boost (see their PR for the result).

The other part is a big refactoring of the way I initialize multi-gpu training with pytorch. The way I wrote it know should be much less bug prone and will make maintenance easier. But I wanted to get that one out of the way before I started fixing some of the feature!
